File: xmlParserJava/firstiter/XMLParser.java
import java.util.LinkedList;



class XML_text_node extends XML_node {
	String text;

	public XML_text_node(String text) {
		this.text = text;
	}
} 

// XML_element_node represents an element

class XML_element_node extends XML_node {
	String name;
	boolean open;

	public XML_element_node(String name) {
		this.name = name;
		this.open = true;
	}

	public XML_element_node() {
		this.name = "root";
	}
} 

class XML_parser {

	// parse_element parses an element beginning with a start tag, followed by
	// zero or more text parts and/or sub-elements, and finally an end tag.
	// It returns an ELEMENT node with subtrees for all text and sub-elements.
	// In case of a parse error or if the start and end tags don't match,
	// NULL is returned.

	XML_element_node parse_element() {

		XML_element_node root = new XML_element_node();

		// System.out.println("\nString to parse: " + s);

		if (s.isEmpty() || s.contains("<>")) {
			return null;
		}

		LinkedList<XML_element_node> visited_elements = new LinkedList<XML_element_node>();

		while (!endofstring) {
			if (currchar == '<') {
				// System.out.println("Found '<' and starting to parse!");
				nextchar();
				// System.out.println("currentChar is this -> " + currchar);
				if (currchar == '/') {
					// System.out.println("Found '/' and starting to parse!");
					nextchar();
					// System.out.println("currentChar is this -> " + currchar);
					boolean element_is_in_list = parse_and_check_closing_tag(visited_elements);
					if (!element_is_in_list) {
						return null;
					}
				} else {
					XML_element_node n = new XML_element_node(parse_element_name());
					// System.out.println("Found this -> " + n.name);
					if (visited_elements.size() == 0) {
						root.subnode = n;
						if (n.name.isEmpty()) {
							return null;
						}
					} else {
						if (visited_elements.getLast().open) {
							if (visited_elements.getLast().subnode != null) {
								visited_elements.getLast().subnode.next = n;

							} else {
								visited_elements.getLast().subnode = n;
							}
						} else {
							visited_elements.getLast().next = n;
						}
					}
					visited_elements.add(n);

				}

			}
			if (currchar == '>') {
				// print_element_pointer();
				// System.out.println("currentChar before nextchar is this -> " + currchar);
				nextchar();
				// System.out.println("currentChar is this -> " + currchar);
				if (Character.isLetter(currchar) || Character.isWhitespace(currchar) && currchar != '\n') {
					XML_text_node t = parse_text();
					// System.out.println("t -> " + t.text);
					// System.out.println("************ " + visited_elements.getLast().name + " is
					// open?: "
					// + visited_elements.getLast().open);
					if (visited_elements.getLast().open) {
						// System.out.println("inserting " + t.text + " as a subnode of " +
						// visited_elements.getLast().name);
						// System.out.println("last visited -> " + visited_elements.getLast().name);
						visited_elements.getLast().subnode = t;
						// System.out.println("inserted t as subnode to " +
						// visited_elements.getLast().name);
						// System.out.println("checking subnode");
						// XML_text_node test = (XML_text_node) visited_elements.getLast().subnode;
						// System.out.println(test.text);
					} else {
						// System.out.println("inserting " + t.text + " as a next of " +
						// visited_elements.getLast().name);
						visited_elements.getLast().next = t;
						// XML_text_node test = (XML_text_node) visited_elements.getLast().next;
						// System.out.println("Testing if it worked! " + test.text);
					}
				} else {
					prevchar();
				}
			}
			// System.out.println("last currchar -> " + currchar);
			if (currchar == '<') {
				prevchar();
			}
			nextchar();
		}

		return (XML_element_node) root.subnode;
	}

	// parse_whitespace will simply skip over any whitespace. Always succeeds.
	boolean parse_whitespace() {
		while (!endofstring && (currchar == ' ' || currchar == '\t' || currchar == '\n')) {
			nextchar();
		}
		return true;
	}

	boolean parse_and_check_closing_tag(LinkedList<XML_element_node> l) {

		String closing_tag = parse_element_name();

		for (XML_element_node node : l) {
			if (node.name.equals(closing_tag)) {
				node.open = false;
				return true;
			}
		}
		return false;
	}

	void print_element_pointer() {

		System.out.println("\n________________");
		System.out.println("string -> " + s);
		System.out.println("currchar -> " + currchar);
		System.out.println("index -> " + p);
		System.out.println("________________");
	}

	// parse_element_name will parse an element name and return a
	// string with the name. Always succeeds unless the name is empty.
	String parse_element_name() {
		String text = "";
		while (!endofstring && Character.isLetter(currchar)) {
			text = text + currchar;
			nextchar();
		}
		return text;
	}

	// parse_text parses text inside elements. It returns a TEXT node with
	// text up to but not including the next tag (or end of string). If there is
	// no such text (the first thing to parse is a tag or end of string,
	// it returns null and leaves s unchanged.

	XML_text_node parse_text() {
		XML_text_node node = null;

		String text = "";
		while (!endofstring && currchar != '<') {
			text = text + currchar;
			nextchar();
		}
		if (!text.equals("")) {
			node = new XML_text_node(text);
		}
		return node;
	}

	// parse_xml will parse a complete XML document string.
	XML_node parse_xml() {
		XML_node out;
		if (parse_whitespace() && (out = parse_element()) != null && parse_whitespace()) {
			if (!endofstring) {
				return null; // Junk after the XML element
			} else {
				return out; // OK.
			}
		} else {
			return null; // Parsing failed.
		}
	}

	String s; // Holds the string being parsed
	int p; // Current position in the string being parsed
	char currchar; // Current character in the string being parsed
	boolean endofstring; // The end of the string being parsed is reached

	// nextchar advances to the next character to be parsed
	void nextchar() {
		p++;
		endofstring = p >= s.length();
		if (!endofstring) {
			currchar = s.charAt(p);
		}
	}

	// nextchar backs up to the previous character in the string to be parsed
	void prevchar() {
		p = p - 2;
		nextchar();
	}

	// Creates a new parser object with a string to be parsed
	XML_parser(String s) {
		this.s = s;
		this.p = -1;
		nextchar();
	}

}

class XML_node {
	XML_node subnode;
	XML_node next;
}

public class YourCode {
}

File: xmlParserC/xml-parser.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <ctype.h>
#include "xml-parser.h"

void print_string_pointer_position()
{
  printf("\n_________________\n");
  printf("Current s -> %s", s);
  printf("\n_________________\n");
}

void ignore_special_chars()
{
  while (!isalpha(*s))
  {
    s++;
  }
}

void push(ioopm_stack_t *s, xml_node_t *pushed_node)
{
  node_t *n = calloc(1, sizeof(node_t));
  n->elem = pushed_node;
  n->next = NULL;

  if (s->size == 0)
  {
    s->top = n;
  }
  else
  {
    n->next = s->top;
    s->top = n;
  }
  s->size++;
}

char *pop(ioopm_stack_t *s)
{
  node_t *old_top = s->top;
  s->top = old_top->next;
  char *poped_value = old_top->elem->text;
  free(old_top);
  s->size--;
  return poped_value;
}

void destroy_stack(ioopm_stack_t *s)
{
  while (s->size != 0)
  {
    pop(s);
  }
  free(s);
}

void print_stack(ioopm_stack_t *s)
{
  node_t *iter = s->top;

  printf("STACK SIZE! -> %d\n", s->size);
  printf("CURRENT STACK!\n");
  while (iter != NULL)
  {
    printf("%s - ", iter->elem->text);
    iter = iter->next;
  }
}

bool element_is_in_stack(ioopm_stack_t *s, char *elem)
{
  node_t *iter = s->top;

  while (iter != NULL)
  {
    //printf("\niter -> %s\n", iter->elem);
    if (strcmp(iter->elem->text, elem) == 0)
    {
      return true;
    }
    iter = iter->next;
  }
  return false;
}

xml_node_t *new_element_node(xml_parse_tree_t *t)
{
  xml_node_t *n = calloc(1, sizeof(xml_node_t));
  n->text = parse_element_name();
  n->type = ELEMENT;
  n->status = OPEN;

  if (t->size == 0)
  {
    t->root = n;
  }
  else
  {
    if (t->current->type == TEXT)
    {
      t->current->next = n;
    }
  }

  t->current = n;
  t->size++;
  //printf("\n current node -> %s, status -> %d\n", t->current->text, t->current->status);
  return n;
}

void find_xml_in_stack_and_close(ioopm_stack_t *s, xml_parse_tree_t *t, char *elem)
{
  node_t *cursor = s->top;
  while (cursor != NULL)
  {
    if (strcmp(cursor->elem->text, elem) == 0)
    {
      cursor->elem->status = CLOSED;
      t->current = cursor->elem;

      return;
    }
    cursor = cursor->next;
  }
}

ioopm_stack_t *create_stack()
{
  ioopm_stack_t *s = calloc(1, sizeof(ioopm_stack_t));
  s->size = 0;
  s->top = NULL;
  return s;
}

xml_parse_tree_t *create_xml_parse_tree()
{
  xml_parse_tree_t *t = calloc(1, sizeof(xml_parse_tree_t));
  t->root = NULL;
  t->current = NULL;
  t->size = 0;
  return t;
}

xml_node_t *parse_element()
{
  ioopm_stack_t *stack = create_stack();
  xml_parse_tree_t *t = create_xml_parse_tree();

  if (strstr(s, "<>"))
  {
    return NULL;
  }

  while (*s)
  {

    if (*s == '<')
    {
      s++;
      if (*s == '/')
      {
        //print_string_pointer_position();
        s++;

        char *closing_tag = parse_element_name();
        //printf("closing_tag -> %s\n", closing_tag);
        if (!element_is_in_stack(stack, closing_tag))
        {
          return NULL;
        }
        find_xml_in_stack_and_close(stack, t, closing_tag);
      }
      else if (isalpha(*s))
      {
        xml_node_t *element_node = new_element_node(t);
        push(stack, element_node);
      }
    }

    if (*s == '>')
    {
      s++;

      if (!isalpha(*s))
      {
        s--;
      }
      else
      {
        xml_node_t *text_node = parse_text();
        if (t->current->status == OPEN)
        {
          //printf("IF OPEN: Current t -> %s\n", t->current->text);
          t->current->subnode = text_node;
          //printf("IF OPEN: Current t subtree -> %s\n", t->current->subnode->text);
        }
        else
        {
          //printf("ELSE CLOSED: Current t -> %s\n", t->current->text);
          t->current->next = text_node;
          //printf("ELSE CLOSED: Current t next -> %s\n", t->current->next->text);
        }

        t->current = text_node;
        t->size++;
        s--;
      }
    }
    s++;
    //printf("\n current node -> %s, status -> %d\n", t->current->text, t->current->status);
  }

  destroy_stack(stack);
  return t->root;
}

// free_xml_tree should take a parse tree and free all allocated memory.

void free_xml_tree(xml_node_t *xml)
{
  free(xml);
}

// parse_whitespace will simply skip over any whitespace. Always succeeds.

bool parse_whitespace()
{
  while (*s == ' ' || *s == '\t' || *s == '\n')
  {
    s++;
  }
  return true;
}

// parse_text parses text inside elements. It should return a TEXT node with
// text up to but not including the next tag (or end of string). If there is
// no such text (the first thing to parse is a tag or end of string,
// it should return NULL and leave s unchanged.

xml_node_t *parse_text()
{
  xml_node_t *node = NULL;
  char *text;
  char *s0;
  s0 = s;
  while (*s != '<' && *s != '\0')
  {
    s++;
  }
  int n = s - s0;
  if (n > 0)
  {
    text = calloc(n + 1, sizeof(char));
    strncpy(text, s0, n);
    node = calloc(1, sizeof(xml_node_t));
    node->type = TEXT;
    node->text = text;
  }
  return node;
}

// parse_element_name will parse an element name and return a newly allocated
// string with the name. If the name is empty, NULL is returned.

char *parse_element_name()
{
  char *text = NULL;
  char *s0;
  s0 = s;
  while (isalpha(*s))
  {
    s++;
  }
  int n = s - s0;
  if (n > 0)
  {
    text = calloc(n + 1, sizeof(char));
    strncpy(text, s0, n);
  }
  return text;
}

// parse_xml will parse a complete XML document string.

xml_node_t *parse_xml(char *in)
{
  xml_node_t *out;
  s = in;
  if (parse_whitespace() &&
      (out = parse_element()) &&
      parse_whitespace())
  {
    if (*s)
    {
      free_xml_tree(out);
      return NULL; // Junk after the XML element
    }
    else
    {
      return out; // OK.
    }
  }
  else
  {
    return NULL; // Parsing failed.
  }
}
